proposals/NNNN-discordbm.md

```diff
@@ -0,0 +1,263 @@
+
+    ...
+
+    public var host: String
+    public var status: HTTPResponseStatus
+    public var version: HTTPVersion
+    public var headers: HTTPHeaders
+    public var body: ByteBuffer?
+
+    ...
+}
+```
+
+This is the basics of `DiscordClient`, but `DefaultDiscordClient` comes with quite a few useful configurations and behaviors as I'll explain next:
+
+```swift
+public struct DefaultDiscordClient: Sendable, DiscordClient {
+    ...
+
+    let client: HTTPClient
+    let configuration: ClientConfiguration
+    let cache: ClientCache
+
+    ...
+}
+```
+
+- The `cache` is a storage for the cached responses.
+  The `configuration` has options to customize how caching is done.
+  `ClientCache` is an actor, and it's shared across all `DefaultDiscordClient`s that have the same authorization token.
+
+The configuration options:
+
+```swift
+public struct ClientConfiguration: Sendable {
+    ...
+
+    public let cachingBehavior: CachingBehavior
+    public var retryPolicy: RetryPolicy?
+    public var performValidations: Bool
+
+    ...
+}
+```
+
+- The `cachingBehavior` can be customized for each endpoint with a different time-to-live.
+  Whenever there is a successful response, the discord-client will cache the response if allowed by this.
+- `performValidations` specifies whether or not to perform client-side validations.
+  This is what `ValidatablePayload` protocol is useful for, which adds a `validate()` func to payloads.
+  The validations only cover what is mentioned in Discord docs. They don't guarantee a valid payload.
+- And there is the `retryPolicy`, which specifies how to retry requests.
+  DiscordBM has some magically-nice behavior built around `retryPolicy`, as I'll explain next.
+
+`DefaultDiscordClient` takes help from a `HTTPRateLimiter` actor to handle rate-limits.
+`HTTPRateLimiter` keeps track of bucket-infos that are available in request headers. Before each request there is a call to `HTTPRateLimiter` that asks "can I do a request now?" and the rate-limiter, considering Discord docs' notes about how exactly Discord rate-limits work, decides to allow a request or not.   
+For example if a bucket is exhausted, the `HTTPRateLimiter` will have no choice but to reject the request, Right?   
+That's not exactly what happens in DiscordBM. If there is a request that is disallowed by the bucket info, `HTTPRateLimiter` instead tries to return the time amount the `DiscordClient` needs to wait before making the request. Then `DiscordClient` takes a look at the `retryPolicy`, and if the `retryPolicy` specifies that requests failed with the `429 Too Many Requests` header can be retried based on headers, `DiscordClient` will just wait the time, and make the request after! This is best of the both worlds:
+
+- Even if you make a lot of requests in a loop, `DiscordClient` not only won't fail, but also won't even let users notice anything. To users it will look like as if they don't even have a rate limit to worry about!
```

Comment:
What if the number of messages per `<timeunit>` exceeds the amount that can be regained at the rate limiter?

Reply:
So to be clear, the whole mechanism doesn't stop you from exceeding your rate-limit.
It makes that happen way less often, but it can't stop you.

I didn't get what you exactly mean, but if you do requests in a loop, it should never fail. What happens is that it for example makes 5 requests immediately, then notices it needs to wait 4.5s before doing the next, and DiscordBM will just wait out that 4.5 before doing the next request, an then continues the loop.

Was that clear? Should i change what is in the proposal?